File: tests/c/flognum1.c
/* Source File: flognum1.c */

#ifdef __CINT__
#include "cint_types.h"
#endif

#ifndef __CINT__
#include <stdio.h>
#include <stdlib.h>
#include <interface.h>
#endif

int main (int argc, char *argv[])
{
  int status;
  int i, counter;
  int numIterations;
  char *cp;
  long value, value1;
  long sub;
  char val[1024];
  char line[1024];

  // check the command-line
  if (argc != 2) {
    printf ("\n\nUsage: %s numIterations\n", argv[0]);
    printf ("\twhere numIterations is the number of times to iterate for each test\n");
    return 1;
  }
  numIterations = atoi (argv[1]);
  if (numIterations <= 0) {
    printf ("\n\n***Error: improper value for numIterations: %d\n", 
	numIterations);
    return _ERROR_;
  }
  // ajust numIterations if it is too low or high
  if (numIterations < 2)
    numIterations = 2;
  if (numIterations > 100)
    numIterations = 100;

  memset (val, 0, 200);

  /*
   * OK, let the user know what the heck is going on and lock and load.
   */
  printf ("\n***Exercise dbnumadd...\n");
  for (i = 1; i <= numIterations; i++) {
    printf ("Iteration %d\n", i);
    cp = dbtestnumber (9, 0);
    strcpy (val, cp);
    strcat (val, ".00");
    value = atol (val);
    counter = 0;
    while (value > 0) {
      counter ++;
      if (counter % 10000 == 0) {
	printf (".");
	fflush (stdout);
      }
      value1 = value;
      cp = dbtestnumber (4, 0);
      sub = atol (cp);
      strcpy (line, "-");
      strcat (line, cp);
      cp = bcnumadd (val, line, 2);
      if (cp == 0) {
	printf ("\n\n***Error: bcnumadd (%s, %s, 2): %s\n",
	    val,line,cdberror);
	bcnumuninit();
	return _ERROR_;
      }
      strcpy (line, cp);
      value = atol (line);
      value1 -= sub;
      sprintf (val, "%ld", value);
      sprintf (line, "%ld", value1);
      status = bcnumcompare (val, line, 2);
      if (status != 0) {
	printf ("%s and %s are not equal and they should be\n",val, line);
      }
    }
    printf ("\ncounter = %d\n", counter);
  }
  // Maybe include the other tests later...or, maybe not.

  bcnumuninit ();

  print_block_list ();
  return 0;
}

File: tests/python/flognum1.py
#!/usr/bin/python2.5

# Load the python module.
import sys
import os
sys.path[0] = ''
os.chdir ("../../lib")
import cdbpy
os.chdir ("../tests/python")

# Lock and load.
if (len (sys.argv) == 2) :
  numIterations = int (sys.argv[1])
  if (numIterations < 2) : numIterations = 2
  if (numIterations > 100) : numIterations = 100
else :
  print "Usage: flognum1.py numIterations"
  sys.exit (1)

print "\n*** Exercise dbnumadd ...\n"
i = 0
while (i < numIterations) :
  i += 1
  sys.stdout.write ("Iteration ")
  sys.stdout.write (str(i))
  sys.stdout.write (" ")
  val = cdbpy.dbtestnumber (10, 0)
  val += ".00"
  counter = 0
  while (float(val) > 0.00) :
    counter += 1
    if (counter % 10000 == 0) :
      sys.stdout.write (".")
      sys.stdout.flush ()
    val1 = val
    num = cdbpy.dbtestnumber (5, 0)
    sub = "-" + num
    val = cdbpy.bcnumadd (val, sub, 2)
    val1f = float (val1) + float (sub)
    val1 = str (val1f)
    status = cdbpy.bcnumcompare (val, val1, 2)
    if (status != 0) :
      print "%s and %s are not equal and they should be" % (val, val1)
  print "\ncounter = %d\n" % (counter)

print "\n*** Exercise dbnumsub ...\n"
i = 0
while (i < numIterations) :
  i += 1
  sys.stdout.write ("Iteration ")
  sys.stdout.write (str(i))
  sys.stdout.write (" ")
  val = cdbpy.dbtestnumber (10, 0)
  val += ".00"
  counter = 0
  while (float(val) > 0.00) :
    counter += 1
    if (counter % 10000 == 0) :
      sys.stdout.write (".")
      sys.stdout.flush ()
    val1 = val
    num = cdbpy.dbtestnumber (5, 0)
    val = cdbpy.bcnumsub (val, num, 2)
    val1f = float (val1) - float (num)
    val1 = str (val1f)
    status = cdbpy.bcnumcompare (val, val1, 2)
    if (status != 0) :
      print "%s and %s are not equal and they should be" % (val, val1)
  print "\ncounter = %d\n" % (counter)

print "\n*** Exercise dbnumdivide ...\n"
i = 0
while (i < numIterations) :
  i += 1
  sys.stdout.write ("Iteration ")
  sys.stdout.write (str(i))
  sys.stdout.write (" ")
  val = cdbpy.dbtestnumber (50, 0)
  val += ".00"
  counter = 0
  while (float(val) > 2.00) :
    counter += 1
    if (counter % 10000 == 0) :
      sys.stdout.write (".")
      sys.stdout.flush ()
    num = cdbpy.dbtestnumber (2, 0)
    val = cdbpy.bcnumdivide (val, num, 2)
  print "\ncounter = %d\n" % (counter)

print "\n*** Exercise dbnummultiply ...\n"
i = 0
while (i < numIterations) :
  i += 1
  sys.stdout.write ("Iteration ")
  sys.stdout.write (str(i))
  sys.stdout.write (" ")
  val = cdbpy.dbtestnumber (2, 0)
  val += ".00"
  topval = cdbpy.dbtestnumber (60, 0)
  topval += ".00"
  counter = 0
  while (float(val) < float (topval)) :
    counter += 1
    if (counter % 10000 == 0) :
      sys.stdout.write (".")
      sys.stdout.flush ()
    num = cdbpy.dbtestnumber (2, 0)
    val = cdbpy.bcnummultiply (val, num, 2)
  print "\ncounter = %d\n" % (counter)

print "\n*** Exercise dbnumraise ...\n"
i = 0
while (i < numIterations) :
  i += 1
  sys.stdout.write ("Iteration ")
  sys.stdout.write (str(i))
  sys.stdout.write (" ")
  val = str (i+1)
  val += ".00"
  topval = cdbpy.dbtestnumber (50, 0)
  topval += ".00"
  counter = 0
  while (float (val) < float (topval)) :
    counter += 1
    if (counter % 10000 == 0) :
      sys.stdout.write (".")
      sys.stdout.flush ()
    num = cdbpy.dbtestnumber (2, 0)
    val = cdbpy.bcnumraise (val, "2.00", 2)
  print "\ncounter = %d\n" % (counter)

cdbpy.bcnumuninit ()
sys.exit (0)
